In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path

# Import confusion_matrix and classification_report from SKLearn
from sklearn.metrics import confusion_matrix, classification_report

# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Import the train_test_learn module
from sklearn.model_selection import train_test_split

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv(
    Path('../Credit_Risk/lending_data.csv')   
)

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns='loan_status')

In [4]:
# Review the y variable Series
y.sample(10)

11886    0
62955    0
3871     0
7698     0
3836     0
41984    0
19909    0
76510    1
10351    0
27276    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:


# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)

Prediction  Actual
73999           0       0
47267           0       0
35950           0       0
42373           0       0
38631           0       0
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [10]:
# Print the classification report for the model
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, predictions, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high-risk loan       0.85      0.91      0.88       619

      accuracy                           0.99     19384
     macro avg       0.92      0.95      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The Healthy loan label is predicted well. 
- The precision and f1-score are 100%. 
- The recall is at .99 showing that some High Risk was predicted as healthy.
- In the High risk loan, the precision, recall and f1-score was slightly low. This is most likely due to a class imbalance between heathy and high risk loans. ( support ratio of 18765 / 619 ).
- The imbalance needs to be addressed to improve the high risk accuracy.

---